### PR TITLE
Add support for `pyproject.toml`

### LINKED ISSUE #3049: Migrate or conjugate `pyproject.toml` with `setup.py`
This issue aims to discuss & integrate `pyproject.toml` into the project introduced in [PEP 518](https://peps.python.org/pep-0518/) to handle builds & pypi packaging.

### PR DESCRIPTION
**Pull request by agriyakhetarpal**
# Description

There are a few nice-to-haves and caveats to this migration to `pyproject.toml`:

1.  The `setup.py` file cannot be removed completely due to its flexibility with custom wheel build processes and steps to build extension modules, so this is more of an addition/conjugation with `setuptools.build_meta` rather than a complete migration—the latter is not possible without the inclusion of a new build-backend such as `meson-build` or `scikit-build-core`. Note: the `scikit-build-core`'s support for editable installations is not mature and might break things, and `scikit-build` is the predecessor to it with limited support for editable installations, which is a deal-breaker.
3. Project metadata can now be included inside `pyproject.toml`, which simplifies the `setuptools.setup()` command. The CMakeBuild.py script is directly included inside `setup.py` to expose the linking of the IDAKLU extension and usage of the CMake arguments.

A list of tasks:

- [x] Editable wheels on Windows, Linux, and macOS are working on GitHub Actions, and on ARM-based macOS machines locally. I will test them on a Windows system locally as well.
- [x] `pip` is able to build non-editable wheels
- [x] Both isolated and non-isolated builds work
- [x] Unit and integration tests are passing on all platforms
- [x] Doctests, example notebooks, and example scripts are passing
- [x] cibuildwheel jobs pass for building reproducible wheels for all platforms

Fixes #3049 and supersedes #3053

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [x] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [x] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Comment by agriyakhetarpal** [comment]
So far unit and integration tests passed for both Windows and macOS but `cibuildwheel` failed on macOS and passed on Windows. With `nox` in the CI we run on a global installation of PyBaMM and not in a virtual environment, so we technically don't _need_ to build without isolation. I am unsure how Read the Docs succeeds—we install the [docs] and [all] extras on an Ubuntu distribution, so it should have failed there like it did for the unit tests and other test groups on Ubuntu...

File: pyproject.toml
Comment (agriyakhetarpal) [inline pyproject.toml:20]: Adding this works, but I am unsure whether this is the right or recommended approach. Adding `cmake` for GNU/Linux helped fix the issues with building editable wheels, allowing all tests to execute—and `casadi` was added as a build-time requirement here to work around the issue of isolated builds not working generally (CMake could not find casadi_DIR otherwise). This way, wheels can now be built both outside and within virtual environments without the need for `--no-build-isolation`. I am not very familiar with the configuration steps (will read up on them in the coming days) for CMakeLists.txt, however, if we can point the correct site-packages directory for `casadi` at the time of isolation, it should fix these issues. @arjxn-py's PR also had a similar issue, but the approach there was to use `setuptools.build_meta:__legacy__`, which is the default build-backend if this configuration value is unspecified.

File: .github/workflows/test_on_push.yml
Comment (agriyakhetarpal) [inline .github/workflows/test_on_push.yml:236]: Now that we have separate jobs for particular test groups (#3264) plus we don't run sessions sequentially either, we don't need this step anymore. The `nox` step will install PyBaMM with its dependencies anyway

File: setup.py
Comment (agriyakhetarpal) [inline setup.py:169]: It is possible that the error with the build isolation came from here, and we can remove `casadi` from the build-time dependencies if this functionality is modified.

Comment (martinjrobins) [reply setup.py:165]: So from what I can see when the wheels are repaired on macos and linux it can't find the casadi lib. In the CMake we execute a bit of python to find the exact directory (in site-packages) that casadi is in ${CASADI_DIR}, then link against this. 

When the wheel is repaired, it needs to search the casadi directory to find the lib, you could either do this by setting LD_LIBRARY_PATHS appropriately, or you could set the RPath for the `idaklu.***.so` file within the CMakeLists.txt file. I would try the latter approach first since you already have the correct CASADI_DIR (since cmake was able to link against the lib in the first place), you need to do something like:

```cmake
target_link_libraries(idaklu PRIVATE ${SUNDIALS_LIBRARIES} casadi) #after this line
set_property(TARGET idaklu APPEND PROPERTY INSTALL_RPATH "${CASADI_DIR}/lib")
```

note, not sure if "${CASADI_DIR}/lib" is even the right subdirectory!

note that I'm not terrible familiar with how to set rpath appropriately, or what cmake does by default, but this is my best guess. If this doesn't work perhaps try setting LD_LIBRARY_DIR appropriately during the fixing wheel step

File: pyproject.toml
Comment (agriyakhetarpal) [reply pyproject.toml]: Resolving this as I think this was/is the right approach: in our case, we are linking against `casadi` at build time for our setup procedure with CMake regardless of whether build isolation is enabled or not. We should not be using the legacy build-backend

File: setup.py
Comment (agriyakhetarpal) [reply setup.py:165]: I fixed it on macOS. I could not get it to work within CMakeLists.txt like you suggested but it turns out that we needed to remove the rpath for `libc++.1.0.dylib` in `libcasadi.3.7.dylib` as well and not only `libcasadi.dylib` (confirmed it with `otool`, both of them had a shared dependency on it). I wonder if we can do the same fixes for GNU/Linux as well because it technically is the same error with `auditwheel` there...

Another solution I thought about (but haven't tested yet, will need to do so in CI) is to override the self-referenced `@rpath` for `libcasadi.3.7.dylib` itself and make it point towards itself with a fixed path, because the rpath will be fixed and installed into `pybamm/.dylibs/` anyway when `cibuildwheel` at the time of repair

**Comment by brosaplanella** [comment]
Shall we close #3053?

**Comment by agriyakhetarpal** [comment]
> Shall we close #3053?

Yes, I think we should. This PR has more progress and is almost complete, I can liaise with @arjxn-py and bring it to fruition.

<hr>

To give an update on the changes here, there are a few more things to be done before I can mark this ready for review. Some of these can also be completed in other, follow-up PRs:

1. The docs build is failing because the text files included inside the package are not being found. They are found on editable installs, but RTD does not do editable installs. To resolve this, I will set up a `MANIFEST.in` file and related configuration to ensure that these files and other files like `CITATIONS.bib` are copied into the SDist (what goes into the source distribution also goes into the wheel). We could use a `setup.cfg` file for this as well, but the Scientific Python guidelines recommend `MANIFEST.in`
2. Resolving the current merge conflicts and fixing the repair procedure on the wheels on GNU/Linux: I managed to do it by fixing the rpath for CasADi's dynamic libraries on macOS, but it is not as simple to do so on GNU/Linux since it uses a Dockerfile from PyPA. I am looking into either modifying the rpath on the `idaklu` target through CMake directly such that it can do so, or setting LD_LIBRARY_PATH to casadi's installation path. @martinjrobins reports that because the newer procedure successfully links against CasADi, delocate and auditwheel should be able to find that path (they do, but are not able to find it for some reason).
3. Moving configuration values for tools like `ruff`, `pre-commit`, `coverage.py`, and `cibuildwheel` into the `pyproject.toml` file such that we can exercise finer-grained control over all of them from the same file.
4. Instead of cloning `pybind11` from GitHub using a script or through `nox`, we can now directly indicate it as a build-time dependency (`pybind11` actually recommends this) in `pyproject.toml` and let CMake find its path through suitable modifications in the `CMakeLists.txt` file.

**Comment by arjxn-py** [comment]
> Shall we close #3053?

Happy to close and move ahead with this. Thanks.

File: .github/workflows/publish_pypi.yml
Comment (agriyakhetarpal) [inline .github/workflows/publish_pypi.yml:29]: `pipx` can get us updated versions of `cibuildwheel`. Otherwise, we could use its action, which is compatible with Dependabot, which has been enabled already

File: .github/workflows/run_periodic_tests.yml
Comment (agriyakhetarpal) [inline .github/workflows/run_periodic_tests.yml:75]: I enabled this step for macOS for this workflow and other ones (turns out we were not building the IDAKLU component in CI since forever!). We were building wheels without ABI or platform tags (`none-any.whl`) otherwise

File: .github/workflows/test_on_push.yml
Comment (agriyakhetarpal) [inline .github/workflows/test_on_push.yml:358]: Now that we have separate jobs for different sets of tests, this step turns out to be redundant (we can save ~1-2 minutes per platform by removing it) because `nox` installs all dependencies in a virtual environment again anyway

File: MANIFEST.in
Comment (agriyakhetarpal) [inline MANIFEST.in:2]: This configuration matches the one with our previous version (23.5). We have historically never included the `tests/` directory and any of the other Markdown files except those inside the `pybamm/` folder. We could, however, provide some more files in the SDist, such as CITATION.cff and other project-specific files

File: noxfile.py
Comment (agriyakhetarpal) [inline noxfile.py:49]: `nox` still doesn't have a `nox.options.verbose` configuration option, however, at least un-silencing fixes the `pip` output

File: setup.py
Comment (agriyakhetarpal) [reply setup.py:164]: Resolving this as I manually appended casadi's installation path to LD_LIBRARY_PATH (I could not fix it through CMake no matter how I tried experimenting with CMAKE_INSTALL_RPATH and BUILD_RPATH) which fixed the issue and pointed `auditwheel` to `libcasadi.so.3.7`

**Comment by agriyakhetarpal** [comment]
I am marking this as ready for review now since wheel builds and the test suite are both passing on all platforms. This PR is slightly big and changed a lot of files, but everything should work now. Note: `casadi` is a build-time dependency in `pyproject.toml` specific to our case because we link against it when compiling the `IDAKLU` target. Now that the wheels repair error is fixed, I now finally realise its cause as to why `auditwheel` and `delocate` failed to find the `casadi` shared objects and dynamic libraries and so files on Ubuntu and macOS: the existence of the PEP517 `pyproject.toml` format mandates the creation of a temporary virtual environment in a private directory when installing a package to establish correct versions of dependencies and avoid conflicts with packages that are already installed. This was not the case with a `setup.py`-only build. I am poised to think that if we were to build and repair the wheels with legacy behaviour (with the use of the `—no-build-isolation` command-line flag and/or the legacy `build_meta` build-backend), the original configuration for setting their rpath would have worked, but we don't need to bother with that anymore unless this solution causes errors sometime in the future, plus we would not want anyone to install without isolation every time. Note: Codacy is not happy with the subprocesses running in the scripts but flags them only when files or lines containing them are changed, perhaps we can ignore those (or suppress them if feasible).

There are some other possible improvements as mentioned in https://github.com/pybamm-team/PyBaMM/pull/3301#issuecomment-1731253161, which I will open separate issues for, for visibility. We should also look into building wheels for ARM-based platforms either natively (Linux runners) or through emulation until GitHub Actions provides native support for macOS (`cibuildwheel` already has support for building universal wheels). The current wheels could be uploaded to TestPyPI in order for us to test them because we aren't checking whether they will install PyBaMM correctly on all platforms currently. They can be downloaded from the job summary page as well

**Comment by agriyakhetarpal** [comment]
@Saransh-cpp this PR has a lot of infrastructure-related changes, it would be great if you could take a look too before the the next release (#3374)

**Review by arjxn-py — COMMENTED** [review]
Thanks @agriyakhetarpal, I tested linux build wheels and they are getting installed for me nicely.

**Comment by agriyakhetarpal** [comment]
I was doing some cleanup in the noxfile to speed up the installation of the optional solvers and it looks like the Windows wheels have started failing unrelated to my changes, while they were passing yesterday and in the other workflow runs in this PR. Is it related to the `vcpkg` issue that has troubled us in the past? I am not sure how to debug the logs. Though it raises a ModuleNotFoundError due to casadi, that is fine here because we don't link against the Python casadi for Windows, we get it from `vcpkg`. Also, it looks like the registry is picking up SUNDIALS `5.7.0`, which is outdated and not the same as our macOS and Linux workflows (we have `6.5.0` there).

File: .github/workflows/publish_pypi.yml
Comment (martinjrobins) [inline .github/workflows/publish_pypi.yml:90]: do we still need this reinstlal of gcc since we install on the line above?

File: .github/workflows/test_on_push.yml
Comment (martinjrobins) [inline .github/workflows/test_on_push.yml:81]: and here

**Review by martinjrobins — APPROVED** [review]
Thanks @agriyakhetarpal, this looks great and vry happy we're migration to a pyproject.toml. Thanks for your work on this, just a question below regarding reinstall of gcc

File: .github/workflows/publish_pypi.yml
Comment (agriyakhetarpal) [reply .github/workflows/publish_pypi.yml:90]: About this: I think it would be wise to keep it. I noticed that three workflow runs on this branch passed without it, but one of them started failing randomly afterwards with the error message that `gfortran` wasn't found, which comes bundled with `gcc`. I think something is up on the GitHub Actions runners' side rather than Homebrew's, but I haven't filed a bug report on either place upstream yet

File: pyproject.toml
Comment (Saransh-cpp) [inline pyproject.toml:13]: Making this [`dynamic`](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#dynamic-metadata) will reduce the number of files in which we have to bump the version for a release -
```
dynamic = ["version"],
```

File: pyproject.toml
Comment (Saransh-cpp) [inline pyproject.toml:181]: See version comment above
```suggestion
[tool.setuptools.packages.find]
include = ["pybamm", "pybamm.*"]

[tool.setuptools.dynamic]
version = {attr = "pybamm.__version__"}
```

File: scripts/update_version.py
Comment (Saransh-cpp) [inline scripts/update_version.py:41]: This can be removed once the version is made dynamic in pyproject.toml

**Review by Saransh-cpp — CHANGES_REQUESTED** [review]
Thanks a ton for working on this, @agriyakhetarpal! Sorry for the delay in the review, but fantastic work! 🎉

File: MANIFEST.in
Comment (agriyakhetarpal) [reply MANIFEST.in:2]: The LICENSE file is included in source distributions by default. I added the CITATION.cff file

File: pyproject.toml
Comment (agriyakhetarpal) [reply pyproject.toml:3]: The PEP660 format was introduced in `setuptools` v64.0.0 (i.e., August 2022), updating it to that (https://stackoverflow.com/a/69711730/14001839)

File: pyproject.toml
Comment (agriyakhetarpal) [reply pyproject.toml:13]: There are a lot of commits in this PR haha but I have tried this. It did not work and some bits of debugging revealed that it was because the `build_meta` build-backend, unlike its legacy counterpart, does not install the optional dependencies first before finding the version as a dynamic attribute. This breaks the PyBaMM import (it cannot import `pybtex` for the Citations class, `anytree`, etc.) and therefore is unable to extract the version string from __version.

I would suggest against doing this for now until we can figure out a way to make the imports truly optional (perhaps by adding warnings that mark limited functionality for a class or feature if a module cannot be imported). This is one of four more issues that I will be opening after this PR is merged (actually three, since I opened #3462 today) to mark as subsequent improvements. The good thing is that we can also look into adding support for `setuptools-scm` now with suitable modifications to the `update_version.py` script.

Comment (Saransh-cpp) [reply pyproject.toml:13]: Ah, sounds good!

**Review by Saransh-cpp — APPROVED** [review]
Looks great, thanks again, @agriyakhetarpal! Could you generate wheels on this branch and share the workflow's link? Other developers can try the wheels locally to ensure everything works before this is merged in. Also, is this blocked by any other PR or is it ready for merging?

**Comment by agriyakhetarpal** [comment]
> Could you generate wheels on this branch and share the workflow's link? Other developers can try the wheels locally to ensure everything works before this is merged in.

I will disable building the wheels on pull requests in a commit at the time just before this will be merged. The latest builds can be downloaded from this link: https://github.com/pybamm-team/PyBaMM/actions/runs/6596974331?pr=3301

@arjxn-py had installed and tried the wheels some while back, so I will re-request a review for testing on amd64 Linux (WSL) and on amd64 (Windows). I will request a review from @kratman so that we can test the wheels on some more Linux machines. We don't have macOS arm64 wheels yet, but perhaps we can tag the maintainer trainees on this PR or on Slack so that anyone who has an Intel-based macOS machine can take a look? I think @tinosulzer could help out too with this. I spun up a container from the latest Docker images under emulation, they were working as expected there.

Note: we don't bundle the functionality or the scripts to run the unit tests in these user installations, so an appropriate and sufficient way to test them might be:

1. To see if `pybamm.have_idaklu()` returns `True`. This would indicate that the IDAKLU extension has been compiled correctly.
2. To see if the example notebooks and/or example scripts pass with PyBaMM installed from the wheel files (just copy `run-tests.py` and the relevant files and run the relevant command). This would allow us to see the IDAKLU solver in usage.

Please let me know if there is a better way to test, although imo these tests are redundant (the source installation in CI builds a wheel in the same way `cibuildwheel` does and therefore we are running the test suite there already). All other dependencies are pure Python and hence do not affect any functionality.

> Also, is this blocked by any other PR or is it ready for merging?

This is ready for merging. Although, @brosaplanella had mentioned in this month's developers' meeting that we would like to wait for it until the release is done, and we now have a pre-release up. While this is a big change, it is related to the packaging only and PyBaMM's functionality will remain the same. So it is up to him and @tinosulzer if this is to be merged before the non-RC, since I can then take a look into some of the further improvements and/or #3462—that would be a nice addition for users considering that the next release (`24.1`) would then come at a much later time, i.e., next year

**Comment by brosaplanella** [comment]
I think the issue was until the RC was published, which it is now. But I will leave it to @Saransh-cpp as he is the one handling the release.

File: noxfile.py
Comment (kratman) [inline noxfile.py:44]:
```suggestion
        session.install("wget", "cmake", silent=False)
```

**Review by kratman — COMMENTED** [review]
Minor review comments. I will not be able to test the functionality until tomorrow.

I will try to test it on MacOS M2 and Linux Mint i5.

File: .github/workflows/publish_pypi.yml
Comment (agriyakhetarpal) [reply .github/workflows/publish_pypi.yml:83]: Makes sense, updating this. I am reinstalling it here because `gcc` is actually available on GitHub Actions runners by default (but sometimes with an erratic configuration of `gfortran`) and Homebrew refuses to install a package if it already exists. `openblas`, `libomp`, and `graphviz` do not come pre-installed so they should be installed rather than reinstalled

File: .github/workflows/publish_pypi.yml
Comment (agriyakhetarpal) [reply .github/workflows/publish_pypi.yml:31]: It can be made a variable in a Lockfile of sorts or in the workflow file itself, but soon we should be able to remove this dependence on cloning a particular branch off the `pybind11` repository and let `pip` fetch the latest stable version. I have plans to do that after this is merged, as noted in [this comment](https://github.com/pybamm-team/PyBaMM/pull/3301?notification_referrer_id=NT_kwDOBG9FzrM3NTU3ODkzMjU2Ojc0NDAxMjMw#issuecomment-1731253161) which was lost in this super long thread haha

**Comment by Saransh-cpp** [comment]
> if this is to be merged before the non-RC

This will/can be merged before the non-RC but into develop and not the release branch. I'd prefer to keep this for the next release.

Only security and bug fixes should be added to the release branch once `rc0` is out.

> if this is to be merged before the non-RC, since I can then take a look into some of the further improvements and/or https://github.com/pybamm-team/PyBaMM/issues/3462—that would be a nice addition for users considering that the next release (24.1) would then come at a much later time, i.e., next year

Is there a specific reason for including this PR in `v23.9` and the further updates in the next one? Both, this PR, as well as other updates can be included in the next release, right?

**Comment by agriyakhetarpal** [comment]
(Note: I thought we were merging `develop` into `main` and adding a git tag on a commit on that branch to create a release. Do we have separate branches for RC and non-RC releases now after #3249, and we push from the RC branch to the non-RC branch to create a stable release? If this is the case, I definitely misunderstood)

> Is there a specific reason for including this PR in v23.9 and the further updates in the next one? Both, this PR, as well as other updates can be included in the next release, right?

No, I don't have any specific reason as such. The follow-up issues and PRs required after this will neither be security/bug fixes as you mention, nor can an estimate be provided on how much time they would take to complete. So I agree that both the changes in this PR plus the follow-ups should make it to the next release, i.e., `v24.1rcX` and subsequently to `v24.1`.

> This will/can be merged before the non-RC but into `develop` and not the release branch. I'd prefer to keep this for the next release.

Yes, this should be good (assuming that the release workflow I wrote about in the above paragraphs is correct).

<hr>

I might have misphrased what I was intending to say, however. To clarify, #3462 is about building wheels for more platforms and can be worked upon as a standalone issue as well. I can start working on that soon: if a PR for it gets completed in time and the wheels work without issues, it would be good for them to make it into the `v23.9` release itself— users on those platforms would then be able to install `v23.9` from `pip` using those wheels and would not have to wait until `the v24.1rcX` or `v24.1` releases, which are many months away at this time. Of course, building wheels is not a security issue or a bug that needs fixing but rather a packaging feature, but it would be a good feature to add in this case. In any case, I will open a PR and try to work it to completion, just that it would be your choice to merge.

**Comment by Saransh-cpp** [comment]
> I thought we were merging develop into main and adding a git tag on a commit on that branch to create a release. 

The [v23.9](https://github.com/pybamm-team/PyBaMM/tree/v23.9) branch will be merged into main to create the final release. Merging PRs in develop will not affect the final release in any way. Any bug/security fix that is required in the final release should be merged into develop (like other PRs) and then cherry-picked to the release branch (`v23.9`).

> Of course, building wheels is not a security issue or a bug that needs fixing but rather a packaging feature

Yes, that sounds like a feature. It would be better to include that in the next release!

**Comment by agriyakhetarpal** [comment]
Makes sense, thanks for the explanation!

**Review by arjxn-py — APPROVED** [review]
I've tested wheels for amd64 Linux (WSL) and amd64 (Windows), they're being installed nicely. I've verified the solvers too by `pybamm.have_idaklu()` and all and also ran unit tests locally. 
Overall tests are running as expected & those with solver dependencies are being skipped.

**Comment by Saransh-cpp** [comment]
@agriyakhetarpal could you merge develop? I was waiting for the release, but now that it is out, we should merge this as soon as possible

**Comment by agriyakhetarpal** [comment]
> @agriyakhetarpal could you merge develop? I was waiting for the release, but now that it is out, we should merge this as soon as possible

Done. Since I am targeting some installation issues, improvements, and documentation in the coming time—it should be fine to merge once @kratman can review (a gentle reminder!)—I can add a minimal test to check for the IDAKLU solver during the `cibuildwheel` steps, but I think it would be better to do that in a follow-up PR since this is already very big.

**Comment by agriyakhetarpal** [comment]
I have disabled the wheel building on PRs like it always was and have triggered a final run on my fork, here is the link: https://github.com/agriyakhetarpal/PyBaMM/actions/runs/6935629522

**Review by kratman — COMMENTED** [review]
Minor fix for the version.

Nox installs/text/examples work on my Apple M2. I will download the wheels and test them as well.

File: pyproject.toml
Comment (agriyakhetarpal) [reply pyproject.toml:13]: Thanks, good catch!

**Comment by agriyakhetarpal** [comment]
> I will download the wheels and test them as well.

@arjxn-py has previously tested ans confirmed on WSL and Windows (amd64). We don't have arm64 wheels yet, so another test on a native Linux machine should be great and would suffice for the most part

**Comment by kratman** [comment]
> > I will download the wheels and test them as well.
> 
> @arjxn-py has previously tested ans confirmed on WSL and Windows (amd64). We don't have arm64 wheels yet, so another test on a native Linux machine should be great and would suffice for the most part

I am testing Linux Mint with x86_64 right now. Should be able to let you know if it works in a couple minutes

**Comment by kratman** [comment]
Works on my Linux machine

**Review by kratman — APPROVED** [review]

**Comment by Saransh-cpp** [comment]
Merging this, I hope everything works fine!

**Comment by agriyakhetarpal** [comment]
So glad to see the purple UI elements now, thanks a ton to everyone who helped me out!

**Pull request merged**
